Optimize compiler phase assembly

retronym opened the pull request:
I noticed in profiles that filtering the the `before` set
for hard links generated a lot of garbage in compilers configured
with a large number of plugins.

This commit filters on the fly instead.

Fixes scala/scala-dev#538

retronym committed:
Optimize compiler phase assembly

I noticed in profiles that filtering the the `before` set
for hard links generated a lot of garbage in compilers configured
with a large number of plugins.

This commit filters on the fly instead.

som-snytt commented:
I'm surprised to hear someone has tricked out their compiler so that they noticed. Or was this just a test scenario?

retronym commented:
This was from Real 🌏 usage, believe it or not.

som-snytt commented on src/compiler/scala/tools/nsc/PhaseAssembly.scala:
```diff
@@ -112,16 +112,19 @@ trait PhaseAssembly {
-          edges -= hl
-          for (edge <- node.before) edge.to = node
+      var befores = node.before
+      def hasHardLinks() = befores.exists(_.hard)
```
This could be a flag `var sawHard` that is set on traversal. However, IIUC (which I probably don't), it has already verified that you can only have one hard dependency on a phase. So just `node.before.find(_.hard)`. You loop if the dependencies you're adding include a hard dependency (at most one).

som-snytt commented on src/compiler/scala/tools/nsc/PhaseAssembly.scala:
```diff
@@ -112,16 +112,19 @@ trait PhaseAssembly {
+        for (hl <- befores) {
+          if (hl.hard) {
+            node.phaseobj = Some(node.phaseobj.get ++ hl.frm.phaseobj.get)
+            node.before = hl.frm.before
```
This doesn't seem right. So IDUC. Doesn't node also have soft dependencies? Namer has a hard dep packageobjects, but maybe many plugins just want to run after namer. Don't answer me, I'll see if I can figure it out.

som-snytt commented on src/compiler/scala/tools/nsc/PhaseAssembly.scala:
```diff
@@ -112,16 +112,19 @@ trait PhaseAssembly {
+            for (edge <- node.before) edge.to = node
+          }
         }
-        hls = Nil ++ node.before.filter(_.hard)
```
Not sure why this archaic idiom is so bad. Is before a large set of soft dependencies? Or making an intermediate one-element hashset just to iterator over it to build a List?

retronym replied:
At a micro-level, its wasteful because it the `Set.filter` creates an intermediate set. `node.before.iterator.filter(_.hard).toList` would be an improvement. But my refactoring was based on the belief that we could just remove the `filter` altogether and just filter out soft links in the lambda argument to `befores.foreach`.

retronym replied:
At least in 2.12, empty `mutable.HashSet`-s aren't that tiny -- they contain a 32-element array for the hash table plus some more data.

som-snytt replied:
Aha.

som-snytt commented on src/compiler/scala/tools/nsc/PhaseAssembly.scala:
```diff
@@ -112,16 +112,19 @@ trait PhaseAssembly {
+      def hasHardLinks() = befores.exists(_.hard)
+      while (hasHardLinks()) {
+        for (hl <- befores) {
+          if (hl.hard) {
```
It's too bad we're scared to write `for (h <- bs if h.hard)`.

som-snytt reviewed (approved):
OK, I see what you did.

som-snytt commented:
The other speculation is at https://github.com/scala/scala/pull/8297

I noticed that the final sort order is by "level" and "name", where the name is the first phase name of a group bound by "runsRightAfter".

So that is why fields runs before tailcalls.

> Node(tailcalls) @ 9 < HashSet() > HashSet(SoftEdge(tailcalls->uncurry)) = Some(List(scala.tools.nsc.Global$tailCalls$@fbaecf85, scala.tools.nsc.Global$specializeTypes$@c48471fb))
> Node(fields) @ 9 < HashSet(SoftEdge(explicitouter->fields)) > HashSet(SoftEdge(fields->uncurry)) = Some(List(scala.tools.nsc.Global$fields$@b4057eb9))

There is a code comment for fields to run before erasure, but it doesn't say runsBefore, which is unused. I wonder why.

One of the partests defines a plugin "ploogin" which happens to precede "terminal" also for alphabetical reasons. This is not reassuring. I wonder if it should warn that the order is arbitrary. Usually you really do care about phase order; not caring is an edge case. (Oh, phase assembly pun!)

diesalbla commented:
@retronym Would this be good enough to satisfy the needs of https://github.com/scala/scala-dev/issues/538 /

som-snytt commented:
Glancing at the linked ticket and its linked WIP, I see that we forgot to ask for a test. And a benchmark. I can't tell if this is deadpan humo(u)r.

retronym committed:
Benchmark for phase assembly

retronym commented:
I'm trying to extract a test case for this. The following, based on 2.13.x, should represent the phase constraints:

```
package scala.tools.nsc

import java.util.concurrent.TimeUnit

import org.openjdk.jmh.annotations._
import org.openjdk.jmh

import scala.tools.nsc.plugins.{Plugin, PluginComponent}

@BenchmarkMode(Array(jmh.annotations.Mode.AverageTime))
@Fork(2)
@Threads(1)
@Warmup(iterations = 5)
@Measurement(iterations = 5)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@State(Scope.Benchmark)
class PhaseAssemblyBenchmark {
  var g: Global = _
  class P[G <: Global with Singleton](g: G) extends Plugin() {
    override val name: String = "plugin"
    override val components: List[PluginComponent] = List(
      new component("phase_0", List("parser"), List("namer")),
      new component("phase_1", List("parser", "phase_0"), List("namer")),
      new component("phase_2", List("typer"), List("refchecks")),
      new component("phase_3", List("phase_1"), List("namer")),
      new component("phase_4", List("typer"), List("phase_7")),
      new component("phase_5", List("typer"), List("phase_6", "phase_13")),
      new component("phase_6", List("typer"), List("phase_11", "phase_13")),
      new component("phase_7", List("pickler", "phase_9"), List("phase_19", "patmat")),
      new component("phase_8", List("typer"), List("phase_9", "phase_13", "phase_11")),
      new component("phase_9", List("typer", "phase_8"), List("superaccessors", "pickler", "phase_17", "phase_13", "phase_11", "phase_7")),
      new component("phase_10", List("typer"), List("refchecks")),
      new component("phase_11", List("typer"), List("phase_14")),
      new component("phase_12", List("typer"), List("explicitouter")),
      new component("phase_13", List("typer"), List("phase_14", "phase_16")),
      new component("phase_14", List("typer"), List("superaccessors")),
      new component("phase_15", List("typer", "phase_14"), List("refchecks")),
      new component("phase_16", List("typer", "phase_14"), List("refchecks")),
      new component("phase_17", List("phase_9", "phase_14", "patmat"), List("phase_18", "phase_19")),
      new component("phase_18", List("phase_16"), List("refchecks")),
      new component("phase_19", List("phase_18", "patmat", "refchecks"), List("explicitouter")),
      new component("phase_20", List("mixin"), List("cleanup")),
      new component("phase_21", List("phase_20"), List("cleanup")),
      new component("phase_22", List("phase_21"), List("cleanup")),
      new component("phase_23", List("delambdafy"), List("jvm"))
    )
    class component(override val phaseName: String, override val runsAfter: List[String], override val runsBefore: List[String]) extends PluginComponent {
      override val global: G = g
      override def newPhase(prev: Phase): Phase = new StdPhase(prev) {
        override def apply(unit: global.CompilationUnit): Unit = ()
      }
    }
    override val description: String = ""
    override val global: G = g
  }

  @Setup
  def setup(): Unit = {
    val global = new Global(new Settings) {
      override def loadPlugins(): List[Plugin] = List(new P[this.type](this))
    }
    global.settings.usejavacp.value = true
    g = global
  }

  @Benchmark def assemble(): Object = {
    g.computePhaseAssembly()
  }
}
```

But allocations for the benchmark are tiny:

```
[info] Benchmark                                                         Mode  Cnt     Score      Error   Units
[info] PhaseAssemblyBenchmark.assemble                                   avgt    5   214.438 ±   24.785   ns/op
[info] PhaseAssemblyBenchmark.assemble:·gc.alloc.rate                    avgt    5  3391.752 ±  382.116  MB/sec
[info] PhaseAssemblyBenchmark.assemble:·gc.alloc.rate.norm               avgt    5  1144.000 ±    0.001    B/op
```

That can also be measured as follows:

```scala

object PhaseAssemblyBenchmark {
  def main(args: Array[String]): Unit = {
    val b = new PhaseAssemblyBenchmark
    b.setup()
    b.assemble()
    b.assemble()
    def allocBytes = java.lang.management.ManagementFactory.getThreadMXBean().asInstanceOf[{ def getThreadAllocatedBytes(id: Long): Long}].getThreadAllocatedBytes(Thread.currentThread.getId)
    val before = allocBytes
    b.assemble()
    println(allocBytes - before)
  }
}
```

Whereas in the real system I'm seeing:

```
scala> :power

scala> def allocBytes = java.lang.management.ManagementFactory.getThreadMXBean().asInstanceOf[{ def getThreadAllocatedBytes(id: Long): Long}].getThreadAllocatedBytes(Thread.currentThread.getId)
allocBytes: Long

scala> g.computePhaseAssembly();g.computePhaseAssembly();{ val before = allocBytes; g.computePhaseAssembly(); allocBytes - before }
49019512

```

I'm still trying to figure out the difference. No rush in merging this PR in the meantime.

retronym commented:
I've pushed a benchmark that shows an improvement under this PR.

## Before

```
[info] Benchmark                                                         Mode  Cnt        Score        Error   Units
[info] PhaseAssemblyBenchmark.assemble                                   avgt   10  1832344.612 ±  56469.975   ns/op
[info] PhaseAssemblyBenchmark.assemble:·gc.alloc.rate                    avgt   10      869.791 ±     27.415  MB/sec
[info] PhaseAssemblyBenchmark.assemble:·gc.alloc.rate.norm               avgt   10  2507836.552 ±     36.410    B/op
[info] PhaseAssemblyBenchmark.assemble:·gc.churn.PS_Eden_Space           avgt   10      875.816 ±    156.800  MB/sec
[info] PhaseAssemblyBenchmark.assemble:·gc.churn.PS_Eden_Space.norm      avgt   10  2523822.122 ± 424947.103    B/op
[info] PhaseAssemblyBenchmark.assemble:·gc.churn.PS_Survivor_Space       avgt   10        0.052 ±      0.074  MB/sec
[info] PhaseAssemblyBenchmark.assemble:·gc.churn.PS_Survivor_Space.norm  avgt   10      150.541 ±    220.173    B/op
[info] PhaseAssemblyBenchmark.assemble:·gc.count                         avgt   10       20.000               counts
[info] PhaseAssemblyBenchmark.assemble:·gc.time                          avgt   10       19.000                   ms
```

## After
```
[info]
[info] Benchmark                                                         Mode  Cnt       Score        Error   Units
[info] PhaseAssemblyBenchmark.assemble                                   avgt   10  908735.258 ±  25198.595   ns/op
[info] PhaseAssemblyBenchmark.assemble:·gc.alloc.rate                    avgt   10     253.003 ±      7.018  MB/sec
[info] PhaseAssemblyBenchmark.assemble:·gc.alloc.rate.norm               avgt   10  362019.469 ±     23.645    B/op
[info] PhaseAssemblyBenchmark.assemble:·gc.churn.PS_Eden_Space           avgt   10     231.316 ±    132.517  MB/sec
[info] PhaseAssemblyBenchmark.assemble:·gc.churn.PS_Eden_Space.norm      avgt   10  330798.077 ± 188745.508    B/op
[info] PhaseAssemblyBenchmark.assemble:·gc.churn.PS_Survivor_Space       avgt   10       0.062 ±      0.140  MB/sec
[info] PhaseAssemblyBenchmark.assemble:·gc.churn.PS_Survivor_Space.norm  avgt   10      88.923 ±    201.403    B/op
[info] PhaseAssemblyBenchmark.assemble:·gc.count                         avgt   10      14.000               counts
[info] PhaseAssemblyBenchmark.assemble:·gc.time                          avgt   10      15.000                   ms
```

som-snytt commented:
Amazeballs. Also, this ethos of attention to numbers has influenced my day job, so thanks for that. Of course, sometimes you wind up cracking open old code you'd rather not touch, oh well.

retronym commented:
Here's how #8297 fares:

```
[info] Benchmark                                                         Mode  Cnt       Score        Error   Units
[info] PhaseAssemblyBenchmark.assemble                                   avgt   10  347619.738 ±  41713.000   ns/op
[info] PhaseAssemblyBenchmark.assemble:·gc.alloc.rate                    avgt   10     653.640 ±    159.663  MB/sec
[info] PhaseAssemblyBenchmark.assemble:·gc.alloc.rate.norm               avgt   10  361560.151 ± 129020.910    B/op
[info] PhaseAssemblyBenchmark.assemble:·gc.churn.PS_Eden_Space           avgt   10     636.441 ±    241.978  MB/sec
[info] PhaseAssemblyBenchmark.assemble:·gc.churn.PS_Eden_Space.norm      avgt   10  351338.909 ± 157772.111    B/op
[info] PhaseAssemblyBenchmark.assemble:·gc.churn.PS_Survivor_Space       avgt   10       0.056 ±      0.063  MB/sec
[info] PhaseAssemblyBenchmark.assemble:·gc.churn.PS_Survivor_Space.norm  avgt   10      30.806 ±     35.903    B/op
[info] PhaseAssemblyBenchmark.assemble:·gc.count                         avgt   10      23.000               counts
[info] PhaseAssemblyBenchmark.assemble:·gc.time                          avgt   10      22.000                   ms
```

retronym commented:
Here's the original documentation about the phase assembly algorithm:

https://www.scala-lang.org/old/sites/default/files/sids/nielsen/Thu,%202009-05-28,%2008:13/compiler-phases-sid.pdf

The performance problem no doubt arises when phases declare more than one soft constraint (ie "runs before phases a, b, and c"). The recursive algorithm branches at these points.

retronym commented:
It feels like we should be using a https://algorithms.tutorialhorizon.com/topological-sort/ algorithm with a twist to deal with the hard links.

retronym commented:
Okay, pushed a benchmark that shows the algorithmic problem. In it, each of the `size` phases declares a dependency to run after the two prior phases.

Before:
```
[info] Benchmark                        (size)  Mode  Cnt        Score       Error  Units
[info] PhaseAssemblyBenchmark.assemble       1  avgt    5     1509.672 ±   450.020  ns/op
[info] PhaseAssemblyBenchmark.assemble       4  avgt    5     5380.669 ±   854.188  ns/op
[info] PhaseAssemblyBenchmark.assemble       8  avgt    5    40450.627 ±  1995.082  ns/op
[info] PhaseAssemblyBenchmark.assemble      16  avgt    5  1255818.095 ± 82024.968  ns/op
```

This PR:
```
[info] Benchmark                        (size)  Mode  Cnt       Score       Error  Units
[info] PhaseAssemblyBenchmark.assemble       1  avgt    5    1003.871 ±    65.471  ns/op
[info] PhaseAssemblyBenchmark.assemble       4  avgt    5    2762.134 ±   193.734  ns/op
[info] PhaseAssemblyBenchmark.assemble       8  avgt    5   13104.495 ±   899.441  ns/op
[info] PhaseAssemblyBenchmark.assemble      16  avgt    5  481946.303 ± 30771.517  ns/op
```

I'll implement a better algorithm which should be proportional to number of phases + number of edges.

som-snytt commented:
Or instead of fsc, fire up fpa for fast phase assembly in the background. Or offload it to the build tool. Obviously the build tool should tell the compiler what pipeline it wants.

retronym committed:
A faster compiler phase assembly algorithm

First, collapse components from hard-linked nodes ("runsRightAfter")
in a pre-processing step.

Then, perform a topological ordering of the graph with a modified
depth first search.

Finally, find the longest path from each node to the initial node
using the result of the topological search to pick the correct
edge. The length the assigns the level.

Existing code compiles the phase list ordering lexicographically by
(level, phaseName).

retronym committed:
Log phase levels under scalac -Xshow-phases -Ydebug -Ylog:all

This shows us the places where we're depending on phase name
as a tie-breaker to decide phase order: fields/tailcalls and
delambdafy/jvm.

```
[log <no phase>] Phase assembly levels: (1,parser)
(2,namer:packageobjects:typer)
(5,patmat)
(6,superaccessors)
(7,extmethods)
(8,pickler)
(9,refchecks)
(10,uncurry)
(11,fields tailcalls:specialize)
(12,explicitouter:erasure:posterasure)
(15,lambdalift)
(16,constructors)
(17,flatten)
(18,mixin)
(19,cleanup)
(20,delambdafy jvm)
(21,terminal)
```

retronym commented:
I've pushed a new implementation:

```
[info] Benchmark                        (size)  Mode  Cnt      Score      Error  Units
[info] PhaseAssemblyBenchmark.assemble       1  avgt    5   2286.511 ±  157.671  ns/op
[info] PhaseAssemblyBenchmark.assemble       4  avgt    5   6555.926 ±  141.870  ns/op
[info] PhaseAssemblyBenchmark.assemble       8  avgt    5  13446.790 ±  801.908  ns/op
[info] PhaseAssemblyBenchmark.assemble      16  avgt    5  33539.428 ± 1036.069  ns/op
```

I've also added debug logging to help find places where we rely on the phase name as a tiebreaker:

```
[log <no phase>] Phase assembly levels: (1,parser)
(2,namer:packageobjects:typer)
(5,patmat)
(6,superaccessors)
(7,extmethods)
(8,pickler)
(9,refchecks)
(10,uncurry)
(11,fields tailcalls:specialize)
(12,explicitouter:erasure:posterasure)
(15,lambdalift)
(16,constructors)
(17,flatten)
(18,mixin)
(19,cleanup)
(20,delambdafy jvm)
(21,terminal)
```

retronym committed:
Be more explicit in our compiler phase ordering

Don't rely on the fact that the assembly algorithm uses the phase
name as a sort key as a tie-breaker.

Before:
```
[log <no phase>] Phase assembly levels: (1,parser)
(2,namer:packageobjects:typer)
(5,patmat)
(6,superaccessors)
(7,extmethods)
(8,pickler)
(9,refchecks)
(10,uncurry)
(11,fields tailcalls:specialize)
(12,explicitouter:erasure:posterasure)
(15,lambdalift)
(16,constructors)
(17,flatten)
(18,mixin)
(19,cleanup)
(20,delambdafy jvm)
(21,terminal)
```

After:

```
[log <no phase>] Phase assembly levels: (1,parser)
(2,namer:packageobjects:typer)
(5,patmat)
(6,superaccessors)
(7,extmethods)
(8,pickler)
(9,refchecks)
(10,uncurry)
(11,fields)
(12,tailcalls:specialize)
(14,explicitouter:erasure:posterasure)
(17,lambdalift)
(18,constructors)
(19,flatten)
(20,mixin)
(21,cleanup)
(22,delambdafy)
(23,jvm)
(24,terminal)
```

som-snytt commented on src/compiler/scala/tools/nsc/PhaseAssembly.scala:
```diff
@@ -166,6 +166,8 @@ trait PhaseAssembly {
         }
         node.level = level
       }
+      def phaseDebug = stack.toSeq.groupBy(_.level).toList.sortBy(_._1).map { case (level, nodes) => (level,  nodes.sortBy(_.phasename).map(_.phaseobj.map(_.map(_.phaseName).mkString(":")).mkString(" ")).mkString(" "))}.mkString("\n")
```
It's definitely time for another Whee! ascii cartoon.

retronym commented:
/synch

retronym commented:
We might be able to do the level-assignment in a more direct (yet still efficient way) with https://en.wikipedia.org/wiki/Bellman%E2%80%93Ford_algorithm modified to find longest path rather than shortest path.
